### PR TITLE
Fix mem leaks on PKCS#12 read error in PKCS12_key_gen_{asc,utf8}

### PR DESCRIPTION
**Pull request by DDvO**
Fixes a leak in each of `PKCS12_key_gen_{asc,utf8}`
~~and a potential leak on multiple iterations within `file_load()` of `loader_file.c`~~.

### CODE REVIEW

**Comment by DDvO** [comment]
Can we please swiftly handle these small fixes as well?
All CI runs have succeeded.

File: crypto/store/loader_file.c [hunk 1]
```diff
@@ -1507,6 +1507,7 @@ static OSSL_STORE_INFO *file_load(OSSL_STORE_LOADER_CTX *ctx,
             unsigned char *data = NULL; /* DER encoded data */
             long len = 0;               /* DER encoded data length */
 
+            OSSL_STORE_INFO_free(result); /* from any previous iteration */
```
Comment (t8m) [inline crypto/store/loader_file.c:1510]: I suppose that result basically cannot be non-NULL here because otherwise the matchcount would be !=0 and we would not loop. @levitte ?

Comment (DDvO) [reply crypto/store/loader_file.c:1510]: I also thought about this and that's why I wrote "potential" on this part.
Yet maybe the `file_load()` function should not rely too much on well-behavior on the functions it calls in between?
At least the inserted `OSSL_STORE_INFO_free()` call does not hurt.
I do not have a witness for an actual mem leak here.

So we could also simply leave out this part of the PR in the interest of a quick inclusion of the other (certain) leak fixes.

Comment (t8m) [reply crypto/store/loader_file.c:1510]: I would then also add `result = NULL;` after the free. Because without it it is also not completely clear a doublefree could not happen.

Comment (t8m) [reply crypto/store/loader_file.c:1510]: Another option would be to add `result == NULL `as the condition for looping.

Comment (DDvO) [reply crypto/store/loader_file.c]: > Another option would be to add `result == NULL `as the condition for looping.

I consider this solution most elegant - done.

Comment (DDvO) [reply crypto/store/loader_file.c]: @t8m, is this fine now?

File: crypto/store/loader_file.c [hunk 2]
```diff
@@ -1564,7 +1564,8 @@ static OSSL_STORE_INFO *file_load(OSSL_STORE_LOADER_CTX *ctx,
             pem_free_flag(pem_name, (ctx->flags & FILE_FLAG_SECMEM) != 0, 0);
             pem_free_flag(pem_header, (ctx->flags & FILE_FLAG_SECMEM) != 0, 0);
             pem_free_flag(data, (ctx->flags & FILE_FLAG_SECMEM) != 0, len);
-        } while (matchcount == 0 && !file_eof(ctx) && !file_error(ctx));
+        } while (result == NULL && matchcount == 0
```
Comment (levitte) [inline crypto/store/loader_file.c:1567]: Why this change?  Has a leak actually been found?  In that case, `file_load_try_decode` should be fixed, as it should *only* return a non-NULL result if `matchcount == 1`

**Review by levitte — CHANGES_REQUESTED** [review]

**Comment by levitte** [comment]
Sorry for the slam in the last minute...  been too busy elsewhere, an this PR only raises questions

File: crypto/store/loader_file.c (hunk 2 above)
Comment (DDvO) [reply crypto/store/loader_file.c]: As mentioned above, I do not have a witness for an actual mem leak at this point.
All right, since this was just a precaution fix for a potential problem and the change is controversal, I've removed it.

File: crypto/pkcs12/p12_key.c [hunk 3]
```diff
@@ -35,8 +35,6 @@ int PKCS12_key_gen_asc(const char *pass, int passlen, unsigned char *salt,
     }
     ret = PKCS12_key_gen_uni(unipass, uniplen, salt, saltlen,
                              id, iter, n, out, md_type);
-    if (ret <= 0)
-        return 0;
     OPENSSL_clear_free(unipass, uniplen);
     return ret;
```
[the remark this answers is not in the document]
Comment (DDvO) [reply crypto/pkcs12/p12_key.c]: @levitte, in this case (and below) the return code is not actually changed because `PKCS12_key_gen_uni()` returns only 0 or 1.
Nevertheless I've adapted the return expression to address your concern and avoid further discussion.

File: crypto/pkcs12/p12_key.c [hunk 4]
```diff
@@ -58,8 +56,6 @@ int PKCS12_key_gen_utf8(const char *pass, int passlen, unsigned char *salt,
...
```
[the remark this answers is not in the document]
Comment (DDvO) [reply crypto/pkcs12/p12_key.c]: Not really - see above.

**Review by levitte — APPROVED** [review]

**Comment by levitte** [comment]
Thanks for your patience.  I'm happy with the result.

**Review by t8m — APPROVED** [review]

**Comment by DDvO** [comment]
Merged, thanks @t8m and @levitte
